lte/gateway/c/core/oai/test/mme_app_task/test_mme_app_emm_encode_decode.cpp

```diff
@@ -0,0 +1,270 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <gtest/gtest.h>
+
+extern "C" {
+#include "lte/gateway/c/core/oai/tasks/nas/emm/msg/AttachRequest.h"
+#include "lte/gateway/c/core/oai/tasks/nas/emm/msg/EmmInformation.h"
+#include "lte/gateway/c/core/oai/tasks/nas/emm/msg/TrackingAreaUpdateRequest.h"
+#include "lte/gateway/c/core/oai/tasks/nas/emm/msg/UplinkNasTransport.h"
+#include "lte/gateway/c/core/oai/common/dynamic_memory_check.h"
+#include "lte/gateway/c/core/oai/common/log.h"
+}
+
+#include "lte/gateway/c/core/oai/test/mme_app_task/mme_app_test_util.h"
+
+namespace magma {
+namespace lte {
+
+#define BUFFER_LEN 200
+
+#define FILL_EMM_COMMON_MANDATORY_DEFAULTS(msg)                                \
+  do {                                                                         \
+    msg.protocoldiscriminator = EPS_SESSION_MANAGEMENT_MESSAGE;                \
+    msg.securityheadertype    = 0x0001;                                        \
+    msg.messagetype           = 33;                                            \
+  } while (0)
+
+class EMMEncodeDecodeTest : public ::testing::Test {
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+ protected:
+  uint8_t temp_buffer[BUFFER_LEN];
+};
+
+TEST_F(EMMEncodeDecodeTest, TestEncodeDecodeAttachRequest) {
+  //   Combined attach, NAS message generated from s1ap tester
+  uint8_t buffer[] = {0x72, 0x08, 0x09, 0x10, 0x10, 0x00, 0x00, 0x00,
+                      0x00, 0x10, 0x02, 0xe0, 0xe0, 0x00, 0x04, 0x02,
+                      0x01, 0xd0, 0x11, 0x40, 0x08, 0x04, 0x02, 0x60,
+                      0x04, 0x00, 0x02, 0x1c, 0x00};
+  uint32_t len     = 29;
+  attach_request_msg attach_request = {0};
+
+  // Decode and encode back message
+  int decoded = decode_attach_request(&attach_request, buffer, len);
+  int encoded = encode_attach_request(&attach_request, temp_buffer, decoded);
+
+  ASSERT_EQ(encoded, decoded);
+  // Check encoded buffer and original buffer are equal
+  EXPECT_ARRAY_EQ(buffer, temp_buffer, encoded);
+
+  bdestroy_wrapper(&attach_request.esmmessagecontainer);
+  bdestroy_wrapper(&attach_request.supportedcodecs);
+}
+
+TEST_F(EMMEncodeDecodeTest, TestDecodeAttachRequestPixel) {
```

Comment:
Can you also include encode_attach_request here, similar to `TestEncodeDecodeAttachRequest`, as this might have more IEs than S1ap tester?